### PR TITLE
fix(emoji-avatar): align sidebar background and support emoji sequences

### PR DESCRIPTION
**Pull request by AtomsH4**
### What this PR does

Before this PR:

- `isEmoji` only matched `\p{Emoji_Presentation}` or `\p{Emoji}️` runs, so ZWJ-joined sequences (`🧛‍♂️`, family emojis `👨‍👩‍👧‍👦`) and keycap emojis (`1️⃣`, `#️⃣`) were misclassified as non-emoji.
- Sidebar `UserAvatar` fell through to the gradient initial-letter branch even when the user's avatar was an emoji — the emoji never rendered.

After this PR:

- `isEmoji` recognizes ZWJ-joined sequences via `EMOJI_SEQUENCE_PATTERN` and keycap emojis via `KEYCAP_EMOJI_PATTERN`. The regex stays at top-of-file as named constants so the intent is self-documenting.
- Sidebar `UserAvatar` renders emoji avatars via the existing `@cherrystudio/ui` `EmojiIcon` (with the same blurred-background treatment used elsewhere in the app), now in a new `fluid` mode that fills the parent's `h-full w-full` instead of a fixed px size.
- `@cherrystudio/ui` `EmojiIcon` gains a `fluid` prop, `aria-hidden="true"` on the blurred background div, and a dedicated primitive-level test. The global `@cherrystudio/ui` renderer mock exposes a stub so consumers can import the primitive in jsdom tests without per-file overrides.

Fixes #

### Why we need it and why it was done in this way

The following tradeoffs were made:

- The emoji regex was rebuilt from `String.raw` fragments instead of a single dense literal. The fragments are slightly slower to read but make the keycap and ZWJ-sequence rules independently auditable.
- Sidebar reuses `EmojiIcon` rather than keeping a near-identical inline copy, accepting `blur-sm` (Tailwind 4 ≈ 4px) instead of the inline `blur-[5px]` for a one-pixel visual difference in exchange for one canonical implementation.
- `fluid` mode is opt-in rather than the default, so the three existing renderer-side `EmojiIcon` consumers (`AssistantAvatar`, `TopicContent`, `AgentLabel`) are unaffected.

The following alternatives were considered:

- A regional-indicator / flag-specific branch — not needed; regional indicators are already `\p{Emoji_Presentation}` and concatenate cleanly via `EMOJI_PART_PATTERN+`.
- Auto-detecting fluid mode by omitting `size` — rejected because the existing `size = 26` default would have flipped into a silent behavior change for callers that omit `size` today.

Links to places where the discussion took place:

### Breaking changes

None — `isEmoji` is more permissive (previously-false sequences now return true, which matches user expectation); Sidebar emoji avatars now render instead of falling through to the gradient initial; `EmojiIcon`'s `fluid` prop defaults to `false`, so existing call sites are unchanged.

### Special notes for your reviewer

- The Sidebar regression test no longer asserts inline class names (`blur-[5px]`, `scale-150`) — it now asserts the emoji is rendered via `EmojiIcon`'s stub and that the gradient-initial branch is not used.
- Originally part of the bundled draft PR #15997 (`fix/ui-regression-polish`), now split out for focused review.

### Checklist

- [x] Branch: This PR targets the correct branch — `main` for active development, `v1` for v1 maintenance fixes
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: Write code that humans can understand and Keep it simple
- [x] Refactor: You have left the code cleaner than you found it (Boy Scout Rule)
- [x] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [x] Documentation: Not required — visual fix and helper update; no user-guide impact.
- [x] Self-review: I have reviewed my own code before requesting review from others

### Release note

```release-note
Fix sidebar user avatars that were blank when the user picked an emoji avatar, and recognize ZWJ-joined and keycap emoji sequences (e.g. 🧛‍♂️, 1️⃣) as emoji throughout the app.
```

### CODE REVIEW

**Review by Pleasurecruise — APPROVED** [review]
Reviewed the PR.

Summary of changes:
- Sidebar user avatars now detect pure emoji values and render them with `EmojiIcon` instead of falling back to gradient initials.
- `EmojiIcon` in `@cherrystudio/ui` now supports a `fluid` mode for parent-sized avatar usage while preserving the default fixed-size behavior.
- `isEmoji` now supports more complete emoji sequences, including ZWJ sequences and keycap emoji.
- Tests were added or updated for Sidebar emoji avatar rendering, `EmojiIcon` fixed/fluid sizing, and emoji detection edge cases.

No blocking issues found. The visible PR checks are passing.

File: src/renderer/utils/naming.ts [hunk 1]
```diff
@@ -139,6 +139,11 @@ export function getLeadingEmoji(str: string): string {
   return match ? match[0] : ''
 }
 
+const EMOJI_PART_PATTERN = String.raw`(?:\p{Emoji}\uFE0F|\p{Emoji_Presentation})`
+const KEYCAP_EMOJI_PATTERN = String.raw`(?:[0-9#*]\uFE0F?\u20E3)`
+const EMOJI_SEQUENCE_PATTERN = String.raw`(?:${EMOJI_PART_PATTERN}(?:\u200D${EMOJI_PART_PATTERN})*)`
+const EMOJI_REGEX = new RegExp(`^(?:${KEYCAP_EMOJI_PATTERN}|${EMOJI_SEQUENCE_PATTERN})+$`, 'u')
```
Comment (eeee0717) [inline src/renderer/utils/naming.ts]: **[B6]** `isEmoji` now recognizes ZWJ/keycap sequences, but `removeLeadingEmoji` and `getLeadingEmoji` above still use the old regex. That leaves the helpers inconsistent: for example, `getLeadingEmoji("🧛‍♂️Assistant")` returns only `🧛`, and `removeLeadingEmoji("🧛‍♂️Alice")` leaves the joiner/gender suffix in the displayed name. Since those helpers feed assistant avatars, chat headers, message anchors, and migrations, please reuse the same sequence/keycap pattern for the leading-emoji helpers and add coverage for ZWJ/keycap inputs there too.

**Review by eeee0717 — COMMENTED** [review]
Found one consistency issue in the emoji helper update.

**Review by MyPrototypeWhat — COMMENTED** [review]
Two small, non-blocking suggestions on the emoji-helper changes. (The `getLeadingEmoji`/`removeLeadingEmoji` consistency point is already covered by the [B6] thread, so I'm not repeating it here.) Nice fix overall.

File: src/renderer/utils/naming.ts (hunk 1 above)
Comment (MyPrototypeWhat) [inline src/renderer/utils/naming.ts]: Pulling these into named constants is a real improvement over the inline regex. The one thing the names still don't convey is what the raw Unicode escapes are — that's the part a future maintainer can't decode at a glance. A one-liner would make the intent fully self-documenting:

```ts
// \uFE0F = VS16 (emoji-presentation selector); \u20E3 = combining enclosing keycap (1️⃣);
// \u200D = ZWJ joining multi-part sequences (🧛‍♂️). Composed into EMOJI_REGEX below.
const EMOJI_PART_PATTERN = String.raw`(?:\p{Emoji}\uFE0F|\p{Emoji_Presentation})`
```

Non-blocking.

File: src/renderer/utils/__tests__/naming.test.ts [hunk 2]
```diff
@@ -85,6 +85,8 @@ describe('naming', () => {
     it('should return true for pure emoji string', () => {
       // 验证纯表情符号字符串返回 true
       expect(isEmoji('😊')).toBe(true)
+      expect(isEmoji('🧛‍♂️')).toBe(true)
+      expect(isEmoji('1️⃣')).toBe(true)
```
Comment (MyPrototypeWhat) [inline src/renderer/utils/__tests__/naming.test.ts:117]: The PR's headline is "support emoji sequences", but the suite locks in only one ZWJ + one keycap case. Since this regex is exactly the kind of code a future refactor is most likely to break, a few more assertions (all green against the current pattern — I verified) would protect the marquee inputs cheaply:

```ts
expect(isEmoji('👨‍👩‍👧‍👦')).toBe(true) // multi-person ZWJ family
expect(isEmoji('🇺🇸')).toBe(true) // regional-indicator flag
expect(isEmoji('👍🏽')).toBe(true) // skin-tone modifier
expect(isEmoji('#️⃣')).toBe(true) // non-digit keycap
expect(isEmoji('😊🌈')).toBe(true) // multi-emoji string
```

Non-blocking.

File: src/renderer/utils/naming.ts (hunk 1 above)
Comment (kangfenmao) [inline src/renderer/utils/naming.ts]: **[B6]** `isEmoji` 现在已经支持 ZWJ/keycap 序列，但 `removeLeadingEmoji` 和 `getLeadingEmoji` 仍然使用旧 regex。实际表现是 `removeLeadingEmoji("🧛‍♂️Alice")` 只移除 `🧛`，会把 `‍♂️Alice` 留在显示名里；`getLeadingEmoji("🧛‍♂️Assistant")` 也只返回半个 emoji。这两个 helper 被 assistant avatar、chat header、message anchor 和 migration 使用，建议复用同一套 sequence/keycap pattern，并为 ZWJ/keycap 的 leading/remove 场景补回归测试。

**Review by kangfenmao — CHANGES_REQUESTED** [review]
需要先修复 emoji helper 的一致性问题：`isEmoji` 已支持 ZWJ/keycap，但 leading/remove helper 仍会截断序列，导致聊天显示名和迁移逻辑留下残余字符。

File: src/renderer/utils/naming.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (AtomsH4) [reply src/renderer/utils/naming.ts]: Fixed in 652557ad3 — `removeLeadingEmoji` and `getLeadingEmoji` now share a start-anchored `EMOJI_LEADING_REGEX` composed from the same `KEYCAP_EMOJI_PATTERN` / `EMOJI_SEQUENCE_PATTERN` parts as `isEmoji`, so ZWJ-joined and keycap sequences are handled consistently. Added regression tests for both helpers.

File: src/renderer/components/Sidebar/__tests__/primitives.test.tsx [hunk 3]
```diff
@@ -12,4 +12,15 @@ describe('UserAvatar', () => {
...
+  it('renders emoji avatars via EmojiIcon (no gradient initials fallback)', () => {
+    const { container } = render(<UserAvatar user={{ name: 'User', avatar: '🌈' }} />)
+
+    const emojiIcon = screen.getByTestId('emoji-icon')
+    expect(emojiIcon).toHaveTextContent('🌈')
+    expect(screen.getByTestId('emoji-icon-background')).toHaveTextContent('🌈')
+    // Emoji avatars must not fall through to the gradient-initial branch.
+    expect(container.firstChild).not.toHaveClass('from-blue-400', 'to-indigo-500')
+    expect(screen.queryByText('U')).not.toBeInTheDocument()
+  })
```
[the remark this answers is not in the document]
Comment (AtomsH4) [reply src/renderer/components/Sidebar/__tests__/primitives.test.tsx:28]: Fixed in cf7a61a88 — the assertion now queries the gradient fallback element directly (`container.querySelector('.from-blue-400')`) instead of the outer wrapper, so it actually fails if the gradient-initial branch renders. The content-based `emoji-icon` assertions are kept.

File: src/renderer/utils/naming.ts (hunk 1 above)
Comment (AtomsH4) [reply src/renderer/utils/naming.ts]: Done in 652557ad3 — both helpers now reuse the shared sequence/keycap pattern via a start-anchored `EMOJI_LEADING_REGEX`, so `getLeadingEmoji('🧛‍♂️Assistant') === '🧛‍♂️'` and `removeLeadingEmoji('🧛‍♂️Alice') === 'Alice'` (no more joiner/gender remnant). Added ZWJ/keycap regression tests for the leading/remove scenarios. Thanks for the catch!

Comment (AtomsH4) [reply src/renderer/utils/naming.ts]: Added in 652557ad3 — the `️` / `⃣` / `‍` documenting comment now sits above the pattern constants. Thanks, agreed this makes the intent fully self-documenting.

File: src/renderer/utils/__tests__/naming.test.ts (hunk 2 above)
Comment (AtomsH4) [reply src/renderer/utils/__tests__/naming.test.ts:117]: Added in 652557ad3 — the suite now also locks in `👨‍👩‍👧‍👦` (family ZWJ), `🇺🇸` (flag), `👍🏽` (skin tone), `#️⃣` (non-digit keycap), and `😊🌈` (multi-emoji), all green against the current pattern. Thanks for pre-verifying them.

File: src/renderer/utils/naming.ts (hunk 1 above)
Comment (AtomsH4) [reply src/renderer/utils/naming.ts]: 已在 652557ad3 修复：`removeLeadingEmoji` / `getLeadingEmoji` 现在复用与 `isEmoji` 相同的 sequence/keycap pattern（新增 start-anchored 的 `EMOJI_LEADING_REGEX`）。现在 `getLeadingEmoji('🧛‍♂️Assistant')` 返回完整的 `🧛‍♂️`，`removeLeadingEmoji('🧛‍♂️Alice')` 返回 `Alice`，不再残留 joiner/gender 后缀。已为 ZWJ/keycap 的 leading/remove 场景补充回归测试。

File: src/renderer/utils/naming.ts [hunk 4]
```diff
@@ -118,14 +118,23 @@ export function firstLetter(str: string): string {
...
 
+// \uFE0F = VS16 (emoji-presentation selector); \u20E3 = combining enclosing keycap (1️⃣);
+// \u200D = ZWJ joining multi-part sequences (🧛‍♂️). Composed into the emoji regexes below.
+const EMOJI_PART_PATTERN = String.raw`(?:\p{Emoji}\uFE0F|\p{Emoji_Presentation})`
+const KEYCAP_EMOJI_PATTERN = String.raw`(?:[0-9#*]\uFE0F?\u20E3)`
+const EMOJI_SEQUENCE_PATTERN = String.raw`(?:${EMOJI_PART_PATTERN}(?:\u200D${EMOJI_PART_PATTERN})*)`
+// Anchored at both ends for whole-string checks (isEmoji); anchored at the start only
+// for leading-run extraction/removal (getLeadingEmoji / removeLeadingEmoji).
+const EMOJI_REGEX = new RegExp(`^(?:${KEYCAP_EMOJI_PATTERN}|${EMOJI_SEQUENCE_PATTERN})+$`, 'u')
+const EMOJI_LEADING_REGEX = new RegExp(`^(?:${KEYCAP_EMOJI_PATTERN}|${EMOJI_SEQUENCE_PATTERN})+`, 'u')
```
Comment (zhangjiadi225) [inline src/renderer/utils/naming.ts]: **[A1]** This new sequence-aware matcher is now shared by `isEmoji`, `getLeadingEmoji`, and `removeLeadingEmoji`, but `firstLetter` directly above still uses the old single-part emoji regex even though its contract says it supports emoji. That leaves the assistant fallback avatar path in `MessageHeader` inconsistent: `firstLetter('🧛‍♂️Bob')` returns only `🧛`, `firstLetter('1️⃣First')` returns `1️`, and `firstLetter('🇺🇸USA')` returns `🇺`. Please make `firstLetter` reuse the same sequence-aware matching logic (or add a first-emoji variant of it before the letter fallback) and cover ZWJ/keycap/flag or skin-tone cases in the `firstLetter` tests.

File: tests/renderer.setup.ts [hunk 5]
```diff
@@ -544,6 +544,13 @@ vi.mock('@cherrystudio/ui', () => {
       React.createElement('div', { ...props, 'data-testid': 'avatar-fallback' }, children),
     EmojiAvatar: ({ children, ...props }) =>
       React.createElement('div', { ...props, 'data-testid': 'emoji-avatar' }, children),
+    EmojiIcon: ({ emoji, className }) =>
```
Comment (zhangjiadi225) [inline tests/renderer.setup.ts]: **[B6]** The Sidebar regression test now proves emoji avatars do not fall back to the gradient branch, but this mock drops the new `fluid` and `fontSize` props. If `UserAvatar` stopped passing `fluid`, the test would still pass even though the real `EmojiIcon` would go back to fixed inline sizing instead of filling the `h-7 w-7` avatar. Please expose those props on the mock, for example via `data-fluid` / `data-font-size`, and assert them in the Sidebar test.

**Review by zhangjiadi225 — CHANGES_REQUESTED** [review]
Found two remaining issues: one runtime consistency gap in firstLetter after introducing the shared emoji sequence matcher, plus one non-blocking test coverage gap around the new EmojiIcon fluid integration.

File: src/renderer/utils/naming.ts (hunk 4 above)
Comment (AtomsH4) [reply src/renderer/utils/naming.ts]: Fixed in dcdb91155. `firstLetter` now uses the shared emoji cluster matcher before the letter fallback, with flag and modifier clusters handled explicitly, so ZWJ/keycap/flag/skin-tone cases return the full leading cluster. Added regression coverage for `🧛‍♂️Bob`, `1️⃣First`, `🇺🇸USA`, and `👍🏽User`.

File: tests/renderer.setup.ts (hunk 5 above)
Comment (AtomsH4) [reply tests/renderer.setup.ts]: Fixed in dcdb91155. The `EmojiIcon` renderer mock now exposes `fluid` and `fontSize` via `data-fluid` / `data-font-size`, and the Sidebar regression test asserts `fluid=true` and `fontSize=10` for emoji avatars.

**Review by zhangjiadi225 — APPROVED** [review]
Both of my points are resolved in dcdb91155 (verified against the latest code):

- **[A1]** `firstLetter` now reuses the shared `EMOJI_CLUSTER_PATTERN` via `FIRST_LETTER_OR_EMOJI_REGEX`, so ZWJ/keycap/flag/skin-tone leading clusters are returned whole — with regression coverage for `🧛‍♂️Bob`, `1️⃣First`, `🇺🇸USA`, `👍🏽User`.
- **[B6]** The renderer mock now exposes `data-fluid`/`data-font-size`, and the Sidebar test asserts `fluid=true` / `fontSize=10`, so a dropped `fluid` prop would now fail the test.

CI is green. LGTM 👍

**Review by MyPrototypeWhat — APPROVED** [review]
